### ISSUE TITLE
As a shopper, I can see my cart is secured during 15 minutes when I start to checkout

### ISSUE THREAD

**Issue by RachL** [openfoodfoundation/wishlist#522]
⚠️ **Please use clockify code #12476 Flower Farms**

## Description
**- As a:** Shopper
**- On page:** `/checkout/details`
**- I want to be able to do:** 

As soon as I start the checkout steps, I can see a little timer telling me how much time my cart is secured. If that time runs out during my checkout process I'm redirected to the cart.
On the cart, the current logic is kept, which means:

- cart are saved for shoppers with accounts as long as the OC stays open
- if items have run out of stock, users are prompted with a message to remove them
- cart are saved for guest checkout as long as they keep browsing on the website. If they leave, the cart is lost

> how much time my cart is secure

This means the products I have in my cart are not available for other shoppers: stock is reserved during 15min process.

If I quit checkout, stock is not secured anymore and I can be redirected to cart if products have become unavailable when I restart checkout later.

Design proposal:

![image](https://github.com/user-attachments/assets/19253755-554e-4353-a72d-594b13ea8b47)

If we do some simple messaging like this ⬆️ we might need to make the menu (or at least the message) sticky on scroll...

## Acceptance Criteria & Tests

To complete once we have agreed on a solution

:warning: this will need a feature toggle / a rollout as it touches a sensitive part of the app, but we will need to turn entire shops to avoid weird conflicts between shoppers of the same shop?

Guest shopper behavior needs to be covered.

## Reference
Raised from
- https://github.com/openfoodfoundation/wishlist/issues/106
- https://community.openfoodnetwork.org/t/secure-products-availability-during-checkout-process/1933/16

**Comment by dacook** [comment]
Partly discussed. Gut feel it could be 5 days develompent, but there is more to consider.
Suggested behaviour (WIP):

* Is the timer active for all shops, or can be enabled on per shop basis?
* Timer is initiated when cart moves into Checkout state (first time you click on checkout)
  *  If you go back to shop page, or navigate to another page: the cart remains in "checkout" state, and continues to appear at top-right of page. So the timer should also appear. Design required (Eg appear as small number within the cart button).
  * Also if you navigate away from shop: the cart remains in top-right
* One timer for all products in cart. Timer is restarted whenever cart is updated
  * add/increment products
  * doesn't need to restart for remove/decrement
  * Also when cart details are updated? (confirm details and payment)
  * What if somebody abuses it and extends their cart for a very long time? Seems unlikely.

**Comment by RachL** [comment]
Thanks @openfoodfoundation/core-devs  !

> Is the timer active for all shops, or can be enabled on per shop basis?

what's the cheapest? :) From product pov I would do it for everyone without the ability to disable it.

> If you go back to shop page, or navigate to another page: the cart remains in "checkout" state, and continues to appear at top-right of page. So the timer should also appear.

It's true that some ecommerce shops do that. I don't think we need to do this here / or for a first version. I would release the stock / stop the timer as soon as the checkout process is left.
Otherwise how would we handle the cart? Currently the cart is kept for shoppers with accounts as long as the OC is open. I find it weird to keep the timer during that time. the process when the timer ends seems also more complex: we would need to push a notification to the user.

> Also if you navigate away from shop: the cart remains in top-right

yes same answer as above, if it's possible technically I would link the timer to the availability of the cart. Otherwise it also means we need to start the timer as soon as the first item lands on the cart?

> add/increment products
> doesn't need to restart for remove/decrement

So yes for me as soon as we leave checkout the timer is reset / the stock in released

> Also when cart details are updated? (confirm details and payment)

IMO, if we can do it easily, I wouldn't update the timer during step 1 and 2 of the checkout.

**Comment by RachL** [comment]
Actually @openfoodfoundation/core-devs sorry for the double ping but after checking this with Theresa, there might be a case where we want the timer to start on the shopfront already (as soon as the first item is put in the cart).
My feeling is this is too big already to be handled on the flower farm budget, so there is a new feature to estimate here: https://github.com/openfoodfoundation/openfoodnetwork/issues/12918

**Comment by David-OFN-CA** [comment]
This is an advanced one i know, but i also know it matters a lot to flower farmers and Theresa - wondering if we could discuss an approach to tackling this via the AI working group? @mkllnk - if I make my first step devising a plan of action (as we discussed last week) and then attach the plan here as an .md file, can we review it and determine if this is too much to bite off or not? or do you or @RachL, ir one of the other devs (I think @dacook did a rough estimate on this already) want to provide me with some bullets on how you'd approach this and what you'd flag as major issues / concerns to be aware of

**Comment by David-OFN-CA** [comment]
@tschumilas if you have any thoughts to share on this item too, please do

**Comment by tschumilas** [comment]
For flower hubs -  the issue is NOT the checkout per se.  The checkout is already much improved.   

The issue is that short supply items are scooped from carts BEFORE checkout.   So we need a timer on the entire shopping experience,  not the checkout.  Customers want items temporarily locked down into their cart BEFORE checkout.  so,  I think this is a more complicated problem.    Is it possible:   Start the timer as soon as the first item goes into the cart.  Timer ends in 15 minutes  or at checkout (whichever is earlier).  Once the timer ends.  items are released back to stock if no checkout occurs.

I know this is quite different from current behaviour,  where an item is still in stock until purchase happens.

I wonder what other's think about this?

**Comment by David-OFN-CA** [comment]
Thanks @tschumilas  - noted! I will hold off on this one, til others have provided feedback

**Comment by mkllnk** [comment]
Yes, before we can dive into a technical solution, we need the right use case. The main conflict is between shoppers. One shopper wants to have the items in their cart secured but other shoppers want that item available to them as well while nobody knows if the first shopper is actually going to buy it.

I'm not aware of any grocery shopping website that reserves stock for cart items. Most online shops don't reserve stock. The only examples I know is when you start the checkout process booking flights. My knowledge may be outdated but the international flights booking system is very old. Any travel agent can reserve seats on a flight from anywhere in the world which can be very helpful but also makes the ticket market a bit volatile and unpredictable.

Anyway, from the shop's perspective, you don't want to reserve stock until you got the money or a definitive sale contract (completed checkout). Otherwise your shop looks pretty empty while you could actually sell more stuff. Holding stock early can also be abused by users.

But in some cases, customers got more frustrated with items going out of stock during shopping. It's a case of regular customers which will buy and don't change their mind and abort the shopping half way through. So this could be an optional feature that the shop can activate.

To start with, we could use a feature toggle to trial this and see what makes sense before offering it to everyone. Later we may want to activate it per shop or for a certain customer group by tag.

The smallest impact would be to reserve stock once you go to the checkout page. But Theresa already said that it's not enough. People want the items in the cart reserved straight away. Technically, we can include line items in carts in the stock calculation. It makes it more expensive but it can happen on-the-fly. We would need to check for line items in carts that are currently edited, maybe with a last_updated_at timestamp within the last 15 minutes. That can then be deducted from the stock level before the stock levels are sent to the browser (JSON API request if I remember correctly).

So, if @tschumilas gives this a thumbs up, and nobody, including AI, finds any serious problems with this approach, you can go ahead with a feature-toggle. Should that toggle be per shop or per customer?

**Comment by RachL** [comment]
> Most online shops don't reserve stock

I have a different view. All shopping experiences that are selling rare product do reserve stock in cart. That's true for transportation (airplane, train tickets), entertainment (cinemas, concert...) but also for short chain food products and craft products.

So you won't find a secured stock-in-cart process on a big supermarket chain that has a lot of stock per item, but you will find it on stores like Etsy or Vinted that manage unique or low stock pieces...just like the OFN.

That being said, I'm slightly nervous we are jumping quickly on a develop solution here. I'm a bit afraid it's going to be a long one to review and test. Is this really the highest priority?

**Comment by tschumilas** [comment]
I know that the primary platform used by flower farmers/hubs in the US (Rooted Farmers)  allows shoppers to 'lock down'an item in their cart (before checking)  for 1 hour.  If the shopper does not checkout by that time,  the item is released back to stock.   Rooted Farmers is only available in the US, But I also know that WooCommerce,  Local Line,  Shopify (other platforms commonly used here) items are not locked down on your order until you complete checkout.   So - it seems there is a mix of approaches out there.

There are 2 top priorities for flower hubs:
1. Advance orders capacity -    that is,  the ability to 'reserve' product from a specific supplier weeks/months in advance,  and then confirm its availability once the time is closer.
2. Timed cart as described here.  

Both are thorny unfortunately.  

If we think timing the shopping experience is too expensive right now,  what about working ["As a customer I can have the option to add products to a previous order in the same OC - #533"](https://github.com/openfoodfoundation/wishlist/issues/533)  .  I know it does not solve the problem of customers accessing short supply products.  But honestly,  the only thing that really solves that problem is more supply.   But if a customer could add to an existing order,  at least they can make repeated purchases,  each locking down product (as they do now).  But this way they would lock that product onto the same order,  so the hub does not have to process different payments.....
